### PR TITLE
Implement "ExpressionLess" compilation engine 

### PR DESCRIPTION
**Pull request by wildmaples**
# Changes

- Implemented the methods required to pass the `ExpressionlessSquare` test files!!! 🎉 
- Refactor how tokens are output and advanced 
- Add some helper methods
- Add the bin script `jack_analyzer` to translate jack files to appropriate XML format

# Next
- Work on expressions with operations (i.e. expressions with multiple terms)

### CODE REVIEW

**Comment by tomstuart** [comment]
I changed this PR to be based on #10 rather than `main` because I think that’s what you intended — I hope that’s right.

File: lib/compilation_engine.rb [hunk 1]
```diff
@@ -57,52 +68,192 @@ def compile_subroutine
...
-    output_token
+    output_token # ;
+
     @output.puts("</varDec>")
   end
 
+  def compile_statements
+    @output.puts("<statements>")
+
+    until !@tokenizer.has_more_tokens? || is_symbol_token_and_equal?("}")
```
Comment (tomstuart) [inline lib/compilation_engine.rb:104]: This condition sticks out as being clumsier and harder to understand than the earlier ones. I remember how we got here, but I think we could now get away with just…

```suggestion
    while @tokenizer.token_type == :KEYWORD
```

…which we could write even more nicely as `while keyword?` (or whatever) if we had a helper for it. (Bonus: checking that `@tokenizer.token_type == :KEYWORD` would also make it more technically correct for us to call `@tokenizer.key_word` on the next line.)

File: lib/compilation_engine.rb [hunk 2]
```diff
@@ -57,52 +68,192 @@ def compile_subroutine
...
-    advance
-    if tokenizer.token_type == :KEYWORD && tokenizer.key_word == :VAR
+    while is_keyword_token_and_equal?(:VAR)
       compile_var_dec
     end
 
-    advance_and_output_token
+    if tokenizer.token_type == :KEYWORD
+      compile_statements
+    end
```
Comment (tomstuart) [inline lib/compilation_engine.rb:238]: In the same way as we removed the conditional around `compile_statements` in `#compile_while` and `#compile_if` (3d538c471b6355cc3d216846979c581137162af2), I think we should do the same thing here:

```suggestion
    compile_statements
```

Again we’d need to update a few unit tests to include the empty `<statements></statements>` tag in the expected output, but I think that’s more consistent with what we’ve done elsewhere.

File: lib/compilation_engine.rb [hunk 3]
```diff
@@ -129,5 +280,14 @@ def output_token
     end
 
     @output.puts("<#{token_type}> #{CGI.escapeHTML(token)} </#{token_type}>")
+    advance
+  end
+
+  def is_symbol_token_and_equal?(symbol)
```
Comment (tomstuart) [inline lib/compilation_engine.rb:286]: Extracting this helper was a great idea. Two nitpicks about its name:

1. It’s too long — it’s fairly obvious what it does so we don’t need to spell out its purpose to this degree.
2. It begins with `is_`, which some style guides [recommend against](https://rubystyle.guide/#bool-methods-prefix) for predicates because it’s redundant and inconsistent with the naming scheme in Ruby’s standard library (e.g. it’s `String#empty?`, not `String#is_empty?`).

So how about something like `#symbol_token?` or even just `#symbol?`?

File: lib/compilation_engine.rb [hunk 4]
```diff
@@ -129,5 +280,14 @@ def output_token
...
+    @tokenizer.token_type == :SYMBOL && @tokenizer.symbol == symbol
+  end
+
+  def is_keyword_token_and_equal?(keyword)
```
Comment (tomstuart) [inline lib/compilation_engine.rb:290]: Ditto: `#keyword_token?` or just `#keyword?` would be better.

Comment (tomstuart) [inline lib/compilation_engine.rb:290]: If the `keyword` argument was optional, or a `*keywords` splat of multiple alternatives, we could use this helper in `#compile_statements` as suggested above.

**Review by tomstuart — APPROVED** [review]
This is excellent progress! Not only are you now parsing some (albeit expressionless) real Jack programs, the parser is now significantly better organised and easier to understand, which I think is going to make it much easier to finish it off. 👏🏻
